feat: TwitterInteractionClient response handling and boredom provider

DanielHighETH opened the pull request:
<!-- Use this template by filling in information and copy and pasting relevant items out of the html comments. -->

# Relates to:
https://x.com/shawmakesmagic/status/1858248520060567911

# Risks
Low


## What does this PR do?


- Added boredom score if conversation is too long or between same people
- Added error handling to the interaction loop using try-catch blocks to ensure the system keeps running smoothly even if an error occurs.
- Improved tweet processing by introducing methods to filter out duplicates (filterValidTweets) and group tweets by conversation (groupTweetsByConversation) for more context-aware responses.
- Made the code more modular by moving timeline formatting (getFormattedTimeline) and response evaluation (evaluateShouldRespond) into their own dedicated methods, making the logic easier to read and reuse.
- Enhanced conversation handling by grouping tweets by conversation ID, enabling the bot to respond with better context for multi-tweet threads.
- Cleaned up logging for better visibility into what the bot is doing and easier tracking of errors.
- Refactored large, complex methods into smaller, single-purpose functions, making the code easier to maintain, debug, and extend in the future.

BugByClaude committed:
Add Korean and French README

MarcoMandar committed:
Merge pull request #23 from ai16z/main

Merge from main

lalalune committed:
Merge pull request #312 from BugByClaude/main

lalalune committed:
Merge pull request #311 from ai16z/shaw/fix-service-calls-maybe

ponderingdemocritus committed:
Merge pull request #299 from thearyanag/269--add-node-version-check-during-build

add node version check

ponderingdemocritus committed:
Merge branch 'main' of https://github.com/ai16z/eliza into starknet-plugin

o-on-x committed:
added working pumpfun.ts

o-on-x committed:
Merge pull request #313 from o-on-x/main

added working pumpfun.ts

ponderingdemocritus committed:
updates

ponderingdemocritus committed:
plugin schema

Commit:
Merge pull request #1 from ai16z/main

Pull Upstream

mcp97 committed:
Merge branch 'main' into llmContentGeneration

alextitonis committed:
Merge branch 'main' into twitter-profile-remake

Commit:
docs: add a new Portuguese README version

madjin committed:
fix broken links in stream notes

madjin committed:
fix docs

lalalune committed:
Merge pull request #321 from ai16z/fix_docs

Fix broken docs

lalalune committed:
Merge pull request #320 from gabrielsants/readme_translate

docs: add a new Portuguese README version

o-on-x committed:
utils.ts example tweet splitting

if no other splitting is done in the other modules would work
interactions.ts & post.ts should then just generate any size no character limit tweet (within reason)
wip

o-on-x committed:
Merge branch 'ai16z:main' into o-on-x-wip-tweet-split

odilitime committed:
be specific about tag/node version, add helpful links, be specific about the characters directory, remove optional sharp since it's installed by default now, run shell => start

odilitime committed:
add collect's fix per madjin, add better sqlite3 rebuild instructions as it keeps coming up

odilitime committed:
Edit the => Create a Character File

MarcoMandar committed:
Merge pull request #24 from ai16z/main

merge from main

MarcoMandar committed:
save trade to the backend

Signed-off-by: MarcoMandar <malicemandar@gmail.com>

madjin committed:
Merge pull request #325 from odilitime/main

Update Quickstart Guide

lalalune committed:
Merge pull request #328 from MarcoMandar/main

Save Trade on creation to the backend

lalalune committed:
Merge pull request #323 from o-on-x/o-on-x-wip-tweet-split

utils.ts example tweet splitting

zo-sol committed:
Update README_KOR.md

The Korean version edited by a Korean

jkbrooks committed:
Merge pull request #309 from Sifchain/llmContentGeneration

[LLM Object Generation][1/2] Leverage AI Lib's Generate Object instead of parsing strings

denizekiz committed:
claude vertex configs added to generation.ts (was missing)

tsubasakong committed:
add Heurist API

tsubasakong committed:
Merge branch 'ai16z:main' into main

lalalune committed:
Merge pull request #330 from denizekiz/claude_vertex

claude vertex configs added to generation.ts (was missing)

lalalune committed:
Merge pull request #329 from sumin13245/patch-1

README_KOR.md Korean version edited by a Korean

lalalune committed:
tsc before statt

lalalune committed:
Merge https://github.com/ai16z/eliza into HEAD

tsubasakong committed:
update Heurist chat models

tsubasakong committed:
update chat models

tsubasakong committed:
add Heurist API key option

tsubasakong committed:
Merge branch 'ai16z:main' into main

lalalune committed:
Merge pull request #293 from pindaroso/main

Dockerized application for local development, testing and deployment

shakkernerd committed:
fix: Build error for packages requiring @ai16z/eliza

- Builds the core package: @ai16z/eliza first
- Excluded the "agent" package from the build process with a specific message.
- Enhanced readability with formatted output for better user experience.

lalalune committed:
Merge pull request #331 from shakkernerd/main

fix: Build error for packages requiring @ai16z/eliza

tsubasakong committed:
Merge branch 'ai16z:main' into main

tsubasakong committed:
clean

tsubasakong committed:
clean log

tsubasakong committed:
back to default model provider

odilitime committed:
clarify testing, add additional docker information

odilitime committed:
formatting, break up copy/paste so easier to run one by one, improve accuracy of character section and how to start it, Add clarify and additional info to common issues

vivoidos committed:
added clientConfig to optionally ignore bots and DMs

lalalune committed:
Merge pull request #333 from odilitime/main

Docs: README.md improvements: clarify testing, add additional docker information

lalalune committed:
Merge pull request #334 from odilitime/fix_quickstart

Docs: additional Quickstart clarification and improvements

lalalune committed:
Merge pull request #336 from vivoidos/client-settings

added clientConfig to optionally ignore bots and DMs

lalalune committed:
Merge pull request #335 from tsubasakong/main

feat: Add Heurist API Integration as New Model Provider

ponderingdemocritus committed:
merge

tsubasakong committed:
add integration docs

tsubasakong committed:
update pepexl model demo settings

tsubasakong committed:
add HEURIST

ponderingdemocritus committed:
fix imports, fix builds

tsubasakong committed:
clean up

ponderingdemocritus committed:
title

lalalune committed:
Merge pull request #287 from ai16z/starknet-plugin

feat: Starknet plugin

mladcepes committed:
Tests setup for starters

ponderingdemocritus committed:
fix dev build

lalalune committed:
Merge pull request #343 from ai16z/fix-dev

fix: dev build

lalalune committed:
Merge pull request #339 from tsubasakong/main

 Update Heurist Integration Documentation and Examples

mladcepes committed:
Adding readme tests file for test documentation

mladcepes committed:
Adding readme tests file for test documentation - Fixing README-TESTS

mladcepes committed:
Removing date check, should think of a better way for generating date

MarcoMandar committed:
Merge pull request #25 from ai16z/main

merge from main

lalalune committed:
Fix some stuff and run prettier

ponderingdemocritus committed:
imports and cleanups

lalalune committed:
Merge branch 'fix-ws' of https://github.com/ai16z/eliza into HEAD

MarcoMandar committed:
getOrCreateRecommenderWithTelegramId

Signed-off-by: MarcoMandar <malicemandar@gmail.com>

lalalune committed:
Merge pull request #345 from MarcoMandar/main

getOrCreateRecommenderWithTelegramId

lalalune committed:
Merge branch 'main' of https://github.com/ai16z/eliza into HEAD

MarcoMandar committed:
Merge pull request #26 from ai16z/main

merge from main

lalalune committed:
some trust fixes

MarcoMandar committed:
Merge remote-tracking branch 'upstream/shaw/trust-fixes'

MarcoMandar committed:
merge-trust-fixes

Signed-off-by: MarcoMandar <malicemandar@gmail.com>

MarcoMandar committed:
calculateBuyAmounts fix and add cache to wallet provider

Signed-off-by: MarcoMandar <malicemandar@gmail.com>

MarcoMandar committed:
update

Signed-off-by: MarcoMandar <malicemandar@gmail.com>

lalalune committed:
Merge pull request #347 from MarcoMandar/main

trust fixes

wjw12 committed:
change default configuration of Heurist

ponderingdemocritus committed:
trust db

ponderingdemocritus committed:
remove

ponderingdemocritus committed:
Merge branch 'main' of https://github.com/ai16z/eliza into trust-db

ponderingdemocritus committed:
clean plugin

ponderingdemocritus committed:
plug

lalalune committed:
Merge pull request #349 from ai16z/trust-db

feat: trust db

thejoven committed:
Add Community & contact and Star History

ponderingdemocritus committed:
fixes

ponderingdemocritus committed:
pnpm

ponderingdemocritus committed:
Merge pull request #353 from thejoven/patch-1

Add Community & contact and Star History

denizekiz committed:
openrouter 70b don't support 128000, changed to 405b in model.ts

lalalune committed:
Merge pull request #354 from ai16z/solana-fixes

fix: solana

lalalune committed:
Merge pull request #346 from ai16z/shaw/trust-fixes

fix: some trust fixes

lalalune committed:
Merge pull request #348 from wjw12/main

change default configuration of Heurist

ponderingdemocritus committed:
add actions

oguzserdar committed:
update tweet interval to 90-180 mins

Temporary solution for controlled post frequency range of 8-16 posts per day

madjin committed:
update docs with new stream notes

bmgalego committed:
fix getRecentMessageInteractions in composeState

lalalune committed:
Merge pull request #360 from oguzserdar/patch-1

update tweet interval to 90-180 mins

mladcepes committed:
Adding tests for relationships, with mocker Users and relationships

ponderingdemocritus committed:
actions

mladcepes committed:
Adding tests for providers.test.ts, with mocking providers and checking negative scenarios

twilwa committed:
Merge pull request #364 from ai16z/new_docs

feat: update docs with new stream notes

mladcepes committed:
Adding messages tests

mladcepes committed:
Merge branch 'main' into feature/tests

ponderingdemocritus committed:
token provider

mladcepes committed:
Runtime file

mladcepes committed:
removing files to sync with main branch

mladcepes committed:
Mocking env file

ai16z-demirix committed:
Merge pull request #1 from ai16z-demirix/feature/tests

Tests setup for starters

mladcepes committed:
Adding not required label to env variables in test documentation

ponderingdemocritus committed:
remove cache

ponderingdemocritus committed:
work

o-on-x committed:
post time set in env

ponderingdemocritus committed:
starknet

o-on-x committed:
Merge pull request #368 from o-on-x/main

post time set in env

o-on-x committed:
default set to new standard  post time 90-180

o-on-x committed:
Merge pull request #369 from o-on-x/main

default set to new standard  post time 90-180 type: post time

ponderingdemocritus committed:
readme

ponderingdemocritus committed:
Merge pull request #370 from ai16z/readme-update

feat: readme

ponderingdemocritus committed:
instal clients from plugin

ponderingdemocritus committed:
Merge pull request #263 from ai16z/twitter-profile-remake

twitter-profile-remake

ponderingdemocritus committed:
Merge pull request #366 from bmgalego/fix-getRecentMessageInteractions

fix: bug in getRecentMessageInteractions not awating for promisses before formating

ponderingdemocritus committed:
token provider

enitrat committed:
feat: starknet token transfer

ponderingdemocritus committed:
Merge branch 'main' of https://github.com/ai16z/eliza into starknet-work

ponderingdemocritus committed:
Merge pull request #373 from enitrat/feat/transfer-token

feat: starknet token transfer

ponderingdemocritus committed:
transfer and cleanup

thearyanag committed:
fix: ca for btc was fake

ponderingdemocritus committed:
fix transfer

tsubasakong committed:
update Heurist image gen settings

oguzserdar committed:
docs: add Turkish README translation (TR)

Part of the internationalization effort to make Eliza more accessible to Turkish-speaking developers.

oguzserdar committed:
docs: add Turkish (TR) README

xclicx committed:
Update README_FR.md

xclicx committed:
Create README_FR.md

ponderingdemocritus committed:
nix

whonion committed:
[docs] add translation README_RU.md

bmgalego committed:
add Groq to getTokenForProvider

ponderingdemocritus committed:
client

ponderingdemocritus committed:
attempt browser

ropresearch committed:
contextual tweet threads and making agents less spammy

DanielHighETH committed:
Boredom score for long convos

ponderingdemocritus committed:
work

DanielHighETH committed:
Refractored TwitterInteractionClient

ponderingdemocritus committed:
fixs client and default character

ponderingdemocritus committed:
Merge pull request #367 from ai16z-demirix/main

feat: Adding unit tests for start - Covering goals, defaultCharacters, relationships, evaulators, posts, database, messages

ponderingdemocritus committed:
Merge pull request #375 from tsubasakong/main

feat: Enhance Heurist Image Generation Settings and Image Handling

ponderingdemocritus committed:
Merge pull request #371 from ai16z/client-plugin

feat: install clients from plugin

ponderingdemocritus committed:
Merge pull request #356 from denizekiz/bugfix/openrouter

fix: openrouter 70b don't support 128000, changed to 405b in model.ts

ponderingdemocritus committed:
improve readme

whonion committed:
Merge branch 'ai16z:main' into main

twilwa committed:
Merge branch 'main' into add-client

twilwa committed:
Merge pull request #382 from ai16z/add-client

feat: client

Junsgh committed:
Fix tweet truncation issue by truncating at complete sentences

wahndo committed:
Update eliza tagline in docusaurus.config.js

ponderingdemocritus committed:
plugin fixes

ponderingdemocritus committed:
revert

ponderingdemocritus committed:
pnpm

ponderingdemocritus committed:
merge

wahndo committed:
docs: fix broken links in agents.md

ponderingdemocritus committed:
nix works

ponderingdemocritus committed:
Merge pull request #355 from ai16z/starknet-work

feat: Complete Starknet DB Trust

wahndo committed:
docs: fix broken links in characterfile.md

wahndo committed:
docs: fix broken links in providers.md

wahndo committed:
docs: added missing section dividers

wahndo committed:
docs: added missing section dividers

wahndo committed:
docs: added missing section dividers

wahndo committed:
docs: added missing section dividers

wahndo committed:
docs: fix broken links, remove toc, revise overview

wahndo committed:
Merge branch 'ai16z:main' into main

ponderingdemocritus committed:
client null

ponderingdemocritus committed:
Revert

ponderingdemocritus committed:
Merge branch 'main' of https://github.com/ai16z/eliza

ponderingdemocritus committed:
Merge pull request #390 from ai16z/fix-client-null

fix: client null

ponderingdemocritus committed:
Merge branch 'main' of https://github.com/ai16z/eliza

wahndo committed:
Merge branch 'ai16z:main' into main

ropresearch committed:
Merge branch 'main' into ropresearch/contextual-threads

lalalune committed:
Merge pull request #374 from thearyanag/fix/ca

fix: ca for btc was spam/fake

boyaloxer committed:
Merge branch 'main' into fix-tweet-truncation

ponderingdemocritus committed:
logging

dorianjanezic committed:
video generation plugin

ponderingdemocritus committed:
revert character

iamladi committed:
Register memory managers if passed to runtime

ponderingdemocritus committed:
Merge pull request #393 from ai16z/logging-improvements

feat: Logging improvements

ponderingdemocritus committed:
unruggable

metadiver committed:
Create README_ES.md

Add spanish translation

lalalune committed:
Merge pull request #388 from boyaloxer/fix-tweet-truncation

Fix tweet truncation issue by truncating at complete sentences

lalalune committed:
Merge pull request #400 from metadiver/main

feat: Create README_ES.md

lalalune committed:
Merge pull request #396 from martincik/fix/register-memory-managers-of-runtime-passed-to-contructor

Register memory managers if passed to runtime

Commit:
correct relative paths DTS Build error #401

ponderingdemocritus committed:
improve logging

ponderingdemocritus committed:
Merge branch 'main' of https://github.com/ai16z/eliza into unruggable

leomercier committed:
Merge pull request #402 from ai16z/fix/tsup-build

fix: tsup build error (client-twitter)

ponderingdemocritus committed:
formatting

ponderingdemocritus committed:
contain

lalalune committed:
Merge pull request #380 from whonion/main

docs: add Russian(RU) translation of README

lalalune committed:
Merge pull request #377 from xclicx/main

docs: Update README_FR.md

lalalune committed:
Merge pull request #376 from oguzserdar/oguzserdar-patch-1

docs: add Turkish (TR) translation of README

ponderingdemocritus committed:
pnpm

ponderingdemocritus committed:
imports

ponderingdemocritus committed:
lock

ponderingdemocritus committed:
Merge pull request #383 from ai16z/ropresearch/contextual-threads

feat: Contextual Twitter Threads + Spam Reduction

ponderingdemocritus committed:
Merge branch 'main' of https://github.com/ai16z/eliza

ponderingdemocritus committed:
lint

ponderingdemocritus committed:
Merge pull request #404 from ai16z/lint

fix: Lint

ponderingdemocritus committed:
console

lalalune committed:
Merge pull request #381 from bmgalego/agent-groq

fix: adds Groq to getTokenForProvider

ponderingdemocritus committed:
blobert

ponderingdemocritus committed:
unruggable

lalalune committed:
Merge pull request #405 from ai16z/twitter-client

fix: console

ponderingdemocritus committed:
pnpm

ponderingdemocritus committed:
Merge pull request #398 from ai16z/unruggable

feat: unruggable

monilpat commented:
Please update the PR to pass the PR Title Check thank you!

monilpat commented on packages/client-twitter/src/interactions.ts:
```diff
@@ -74,288 +74,234 @@ IMPORTANT: {{agentName}} (aka @{{twitterUserName}}) is particularly sensitive ab
 
     async handleTwitterInteractions() {
-        console.log("Checking Twitter interactions");
+        console.log("Checking Twitter interactions...");
```
Let's remove these console.log statements before merging in!

monilpat commented on packages/client-twitter/src/interactions.ts:
```diff
@@ -74,288 +74,234 @@ IMPORTANT: {{agentName}} (aka @{{twitterUserName}}) is particularly sensitive ab
+    filterValidTweets(tweets: Tweet[]): Tweet[] {
+        const seenIds = new Set();
+        return tweets
+            .filter((tweet) => tweet.userId !== this.twitterUserId) // Exclude bot's tweets
+            .filter((tweet) => tweet.text && !seenIds.has(tweet.id)) // Exclude empty or duplicate tweets
```
We can combine these filters into one filter that ||s the two checks to improve performance

DanielHighETH replied:
Combined

monilpat commented on packages/client-twitter/src/interactions.ts:
```diff
@@ -74,288 +74,234 @@ IMPORTANT: {{agentName}} (aka @{{twitterUserName}}) is particularly sensitive ab
+    }
+
+    async handleConversation(conversationId: string, tweets: Tweet[]) {
+        console.log(`Processing conversation: ${conversationId}`);
```
same here

DanielHighETH replied:
All console.logs removed

lalalune committed:
update readme

lalalune committed:
Merge branch 'main' of https://github.com/ai16z/eliza into HEAD

monilpat commented on packages/plugin-bootstrap/src/providers/boredom.ts:
```diff
@@ -326,6 +330,28 @@ const boredomProvider: Provider = {
+        const recentUsers = Array.from(recentUserActivity);
+
+        // penalty for repetitive recent interactions
+        // probably change totalMessages value
```
Can we address this comment or add a TODO with an issue #? Thank you!

monilpat commented on packages/plugin-bootstrap/src/providers/boredom.ts:
```diff
@@ -336,8 +362,6 @@ const boredomProvider: Provider = {
         );
         const selectedMessage = boredomLevel.statusMessages[randomIndex];
         return selectedMessage.replace("{{agentName}}", agentName);
```
Thanks for cleaning this up!

ponderingdemocritus committed:
fixes bugs

ponderingdemocritus committed:
blobert

ponderingdemocritus committed:
fixe

DanielHighETH committed:
removed console logs

ponderingdemocritus committed:
Merge pull request #407 from ai16z/fixes

fix: Fixes

DanielHighETH committed:
Filters combined to one

mcp97 reviewed (changes requested):
LGTM need some minor clean up please address then will approve!

DanielHighETH committed:
include TODO comments for #408

Added TODO comments for #408 to review and update the `totalMessages` logic.

DanielHighETH commented:
> LGTM need some minor clean up please address then will approve!

Everything commited and issue #408 created! Thank you

wahndo committed:
docs: capitalize flexible in tagline

monilpat commented:
Amazing - thank you so much!

mcp97 reviewed (dismissed):
LGTM after addressing comments!

jkbrooks committed:
Merge pull request #389 from wahndo/main

docs: refresh eliza's tagline, fix broken links, unify formatting for core concepts

jkbrooks commented:
@monilpat you sure? we have conflicts in `packages/client-twitter/src/interactions.ts`

monilpat commented:
I can't see the conflicts somehow. That is so weird. Yes - please resolve the conflicts first. Sending back to reviewer.

jkbrooks commented:
@DanielHighETH would you fix the merge conflict in `packages/client-twitter/src/interactions.ts` so we can finish the review and merge this in?

whalelephant committed:
docs: add GROK_API_KEY

DanielHighETH commented:
> @DanielHighETH would you fix the merge conflict in `packages/client-twitter/src/interactions.ts` so we can finish the review and merge this in?

@jkbrooks @monilpat a new code was merged 3 hours ago from https://github.com/ai16z/eliza/pull/383 I'm not really sure if we should use my refractored code in this new code

lalalune committed:
Merge pull request #409 from whalelephant/main

docs: add GROK_API_KEY

fabrizioff committed:
docs: add italian README translation

ponderingdemocritus committed:
services

ponderingdemocritus committed:
remove character

ponderingdemocritus committed:
fix logging

ponderingdemocritus committed:
linting

o-on-x committed:
removed ollama embeddings. fastembeddings or openai only

jkbrooks committed:
Merge pull request #411 from fabrizioff/docs/translate-readme-it

Add italian README.md translation

dorianjanezic committed:
Add video generation plugin tests and constants

ponderingdemocritus committed:
systems work

ponderingdemocritus committed:
character

ponderingdemocritus committed:
services

sirkitree committed:
Merge pull request #394 from dorianjanezic/main

feat: video generation plugin

denizekiz committed:
Readme update WSL  2 link added. (#419)

* WSL 2 instrunctions link added to readme.md

* WSL2 instructions to readme.md added

ponderingdemocritus committed:
merge

ponderingdemocritus committed:
issues

sirkitree committed:
Merge pull request #412 from ai16z/fix-services

feat: services

sirkitree committed:
Merge pull request #413 from o-on-x/main

fix: removed ollama embeddings. fastembeddings or openai only

jkbrooks commented:
@DanielHighETH  looks like there is still some value in your PR, even though we also have #383 

Please let us know if you think you can get to the merge conflict, we are still happy to merge. Otherwise, no worries, we can close this PR.

DanielHighETH commented:
@jkbrooks is there a way to DM you, for some reason I can't pull from the original repo, so I can't resolve the conflicts...

DanielHighETH committed:
refractored newest commit

DanielHighETH committed:
conflicts resolved

DanielHighETH committed:
Merge branch 'main' into main

monilpat commented:
Thanks for doing this and addressing all comments. Sorry to bother you again, but can we add any tests or screengrabs confirming that this does in fact improve the conversation and address? https://x.com/shawmakesmagic/status/1858248520060567911

DanielHighETH commented:
> Thanks for doing this and addressing all comments. Sorry to bother you again, but can we add any tests or screengrabs confirming that this does in fact improve the conversation and address? https://x.com/shawmakesmagic/status/1858248520060567911

Hey, I have no idea how to do it actually
